src/client/ua_client_connect.c

```diff
@@ -409,6 +409,25 @@ sendHELMessage(UA_Client *client) {
     return UA_STATUSCODE_GOOD;
 }
 
+void processRHEMessage(UA_Client *client, const UA_ByteString *chunk) {
+    UA_LOG_INFO(&client->config.logger, UA_LOGCATEGORY_CLIENT, "RHE received");
+
+    size_t offset = 0; /* Go to the beginning of the TcpHelloMessage */
+    UA_TcpReverseHelloMessage rheMessage;
+    UA_StatusCode retval =
+        UA_decodeBinaryInternal(chunk, &offset, &rheMessage,
+                                &UA_TRANSPORT[UA_TRANSPORT_TCPREVERSEHELLOMESSAGE], NULL);
+
+    if(retval == UA_STATUSCODE_GOOD) {
+        UA_String_clear(&client->endpointUrl);
+        UA_String_copy(&rheMessage.endpointUrl, &client->endpointUrl);
+    }
+
+    UA_TcpReverseHelloMessage_clear(&rheMessage);
+
+    sendHELMessage(client);
```

Comment:
sendHELMessage will set the state to `UA_SECURECHANNELSTATE_HEL_SENT`.
To me that is a different state. As in "we have sent HEL, but have not receive an ACK.

I'd rather set to `UA_SECURECHANNELSTATE_ACK_RECEIVED` after this sendHelMessage succeeds.
As in "the initial handshake is done".

Maybe I don't understand the logic.
Isn't the client supposed to send the following OPN request to open a SecureChannel?
That is only done by the client if the state is `UA_SECURECHANNELSTATE_ACK_RECEIVED`.

Reply:
@jpfr The server's RHE message is the first message that is sent over the TCP connection and must be answered with a HEL message by the client (see https://reference.opcfoundation.org/Core/Part6/v104/docs/7.1.3#_Ref456138622). After this, the sequence is the same as for a normal connect.
I like the idea of having this intermediate UA_SECURECHANNELSTATE_HEL_SENT state so I can see where the process got stuck if the ACK message is not sent due to some error.